File: Codes in Python3/Recursion&BackTracking/M-Coloring.py
from typing import *
def isValid(node:int,mat: List[List[int]], color: List[int], n: int,col: int) -> bool:
    for i in range(0,n):
        if i!=node and mat[i][node]==1 and color[i]==col:
            return False
    return True

def solve(node:int,mat: List[List[int]], color: List[int], n: int,m: int) -> bool:
    if node==n:
        return True
    for i in range(1,m+1):
        if isValid(node,mat,color,n,i):
            color[node]=i
            if solve(node+1,mat,color,n,m):
                return True
            color[node]=0
    return False


def graphColoring(mat: List[List[int]], m: int) -> str:
    n=len(mat)
    color = [0]*n
    if solve(0,mat,color,n,m):
        return "YES"
    else:
        return "NO"
    pass 

File: Codes in C++/Recursion&BackTracking/M-Coloring.cpp
bool isValid(int node,vector<vector<int>> &mat,vector<int> &color,int col)
{
	for (int i = 0; i < mat.size(); i++)
	{
		if (i!=node && mat[node][i] && color[i] == col)
		{
			return false;
		}
	}

	return true;
}

bool solve(int node,vector<vector<int>> &mat, vector<int> &color ,int n,int m)
{
    if(node==n)
        return true;

    for(int i=1;i<=m;i++)
    {
        if(isValid(node,mat,color,i))
        {
            color[node]=i;
            if(solve(node+1,mat,color,n,m))
            return true;
            color[node]=0;
        }
    }

    return false;
}

string graphColoring(vector<vector<int>> &mat, int m) {
    int n=mat.size();
    vector<int> color(n,0);
    if(solve(0,mat,color,n,m)) return "YES";
    return "NO";
}